RavenDB-14435 Don't stops replication after double dispose

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14435 Replication stops after an exception

karmeli87 committed:
RavenDB-14435 Prevent double dispose of incoming replication handler

ppekrol commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1096,8 +1097,13 @@ private void AddReplicationPerformance(IncomingReplicationStatsAggregator stats)
         public void Dispose()
         {
+            if (_disposed.Raise() == false)
+                return; // already disposed
```
Wouldn't it be better to use DisposeOnce? It was created for this purpose.

karmeli87 replied:
done

ppekrol commented on src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs:
```diff
@@ -357,7 +357,7 @@ private void Replicate()
                     {
                         _parent.DebugWaitAndRunReplicationOnce.Wait(_cts.Token);
-                        _parent.DebugWaitAndRunReplicationOnce.Reset();
+                        _parent.DebugWaitAndRunReplicationOnce?.Reset();
```
We checked that this is not null 2 lines up?

karmeli87 replied:
in the test I do
```
database.ReplicationLoader.DebugWaitAndRunReplicationOnce = null;
mre.Set();
```

karmeli87 committed:
RavenDB-14435 address PR comments

- use dispose once.
- avoid race between adding and removing incoming connection, and simply replace the connection if it is disposed.
